pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,488 @@
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	mapiv1beta1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	machineAnnotationKey   = "machine.openshift.io/machine"
+	machineInternalIPIndex = "machineInternalIPIndex"
+	machineProviderIDIndex = "machineProviderIDIndex"
+	nodeInternalIPIndex    = "nodeInternalIPIndex"
+	nodeProviderIDIndex    = "nodeProviderIDIndex"
+)
+
+// blank assignment to verify that ReconcileNodeLink implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNodeLink{}
+
+// ReconcileNodeLink reconciles a Node object
+type ReconcileNodeLink struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	// This is useful for unit testing so we can mock cache IndexField
+	// and emulate Client.List.MatchingField behaviour
+	listNodesByFieldFunc    func(key, value string) ([]corev1.Node, error)
+	listMachinesByFieldFunc func(key, value string) ([]mapiv1beta1.Machine, error)
+}
+
+// Add creates a new Nodelink Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	reconciler, err := newReconciler(mgr)
+	if err != nil {
+		return fmt.Errorf("error building reconciler: %v", err)
+	}
+	return add(mgr, reconciler, reconciler.nodeRequestFromMachine)
+}
+
+func indexNodeByProviderID(object runtime.Object) []string {
+	if node, ok := object.(*corev1.Node); ok {
+		if node.Spec.ProviderID != "" {
+			klog.V(3).Infof("Adding providerID %q for node %q to indexer", node.Spec.ProviderID, node.GetName())
+			return []string{node.Spec.ProviderID}
+		}
+		return []string{}
+	}
+	klog.Warningf("Expected a node for indexing field, got: %T", object)
+	return []string{}
+}
+
+func indexMachineByProvider(object runtime.Object) []string {
+	if machine, ok := object.(*mapiv1beta1.Machine); ok {
+		if machine.Spec.ProviderID != nil {
+			if *machine.Spec.ProviderID != "" {
+				klog.V(3).Infof("Adding providerID %q for machine %q to indexer", *machine.Spec.ProviderID, machine.GetName())
+				return []string{*machine.Spec.ProviderID}
+			}
+		}
+		return []string{}
+	}
+	klog.Warningf("Expected a machine for indexing field, got: %T", object)
+	return []string{}
+}
+
+func indexNodeByInternalIP(object runtime.Object) []string {
+	var keys = []string{}
+	node, ok := object.(*corev1.Node)
+	if !ok {
+		klog.Warningf("expected a node for indexing field, got: %T", object)
+		return keys
```

Comment:
We should be explicit with the return type and only return non-nil when we actually have some keys: 

```go
func indexNodeByInternalIP(object runtime.Object) []string {
	node, ok := object.(*corev1.Node)
	if !ok {
		klog.Warningf("expected a node for indexing field, got: %T", object)
		return nil
	}

	var keys = []string

	for _, a := range node.Status.Addresses {
		if a.Type == corev1.NodeInternalIP {
			keys = append(keys, a.Address)
			klog.V(3).Infof("Adding internal IP %q for node %q to indexer", a.Address, node.GetName())
		}
	}

	return keys
}
```

Reply:
And the same comment https://github.com/openshift/machine-api-operator/pull/321/commits/d687214415f30eb550cc75daa2b5325b677b47e4#r291497921 applies to the other indexing functions.